### PR TITLE
Fix #1663 Typing - Use explicit optional

### LINKED ISSUE #1663: Remove usage of implicit Optional
> A past version of this PEP allowed type checkers to assume an optional type when the default value is None, as in this code:
> 
> def handle_employee(e: Employee = None): ...
> 
> This would have been treated as equivalent to:
> 
> def handle_employee(e: Optional[Employee] = None) -> None: ...
> 
> This is no longer the recommended behavior. Type checkers should move towards requiring the optional type to be made explicit.

https://www.python.org/dev/peps/pep-0484/

We currently use implicit `Optionals`. mypy still allows them but may change this in the future.

### PR DESCRIPTION
**Pull request by nadege**
To fix #1663
Also added the mypy option to check for implicit optional.

### CODE REVIEW

File: src/marshmallow/base.py [hunk 1]
```diff
@@ -31,16 +32,29 @@ def _deserialize(self, value, attr, data, **kwargs):
 class SchemaABC:
     """Abstract base class from which all Schemas inherit."""
 
-    def dump(self, obj, *, many: bool = None):
+    def dump(self, obj, *, many: typing.Optional[typing.Optional[bool]] = None):
```
Comment (sloria) [inline src/marshmallow/base.py]:
```suggestion
    def dump(self, obj, *, many: typing.Optional[bool] = None):
```

Here and elsewhere

**Review by sloria — COMMENTED** [review]
Thanks for the PR. Can you fix the "doubly-wrapped" `Optional` usages here?

File: src/marshmallow/base.py (hunk 1 above)
Comment (nadege) [reply src/marshmallow/base.py]: oops, I'll do that :)

Comment (nadege) [reply src/marshmallow/base.py]: fixed

**Comment by lafrech** [comment]
Thanks @nadege. Looks good. Do you mind rebasing (conflict in AUTHORS.rst file)?

**Comment by nadege** [comment]
rebased :)

**Pull request merged**